### PR TITLE
PMP: Limited Number of Self-intersections 

### LINKED ISSUE #5828: Add a version of self-intersections that reports only a limited number of SI
In PMP, the function `self_intersections` reports all pairs of intersecting faces of a given mesh. If there are a LOT of self-intersections a user might not be interested in getting all the pairs (for performance reason). We should try to add a named parameter to that function that will report self-intersections up to a given number provided by the user.

A concurrent version of `self_intersections()` was recently introduced. Basically, face pairs are collected in a concurrent vector before being put in the user output iterator. So in that case using a filter output iterator is not possible (well possible but without any speed up). Note that it is fine if in the concurrent version, the number of self-intersection reported is not exactly the number expected but at least the threshold requested (should be doc).

### PR DESCRIPTION
**Pull request by maxGimeno**
## Summary of Changes
Add a named-parameter for limiting the number of SI computed.

## Release Management

* Affected package(s):PMP
* Issue(s) solved (if any): fix #5828 
* Feature/Small Feature (if any): [Limited_Number_of_Self_Intersections](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Limited_Number_of_Self_Intersections) -- pre-approved **2021/07/09**
* Link to compiled documentation (obligatory for small feature) [self_intersections() v1](https://cgal.github.io/5837/v1/Polygon_mesh_processing/group__PMP__intersection__grp.html#gaf19c80ec12cbff7ebe9e69453f1d40b8)
* License and copyright ownership: GPL-3+

### CODE REVIEW

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h [hunk 1]
```diff
@@ -290,14 +353,39 @@ self_intersections_impl(const FaceRange& face_range,
...
+        std::atomic<unsigned int> atomic_counter(counter);
+        Throw_functor throwing_count_functor(atomic_counter, max_number, std::back_inserter(face_pairs));
+        Throwing_after_count_output_iterator count_filter(throwing_count_functor);
+        Filtered_intersecting_faces_filter limited_callback(tmesh, vpmap, gt, count_filter);
+        CGAL::box_self_intersection_d<ConcurrencyTag>(box_ptr.begin(), box_ptr.end(), limited_callback, cutoff);
+      }
+      catch(const CGAL::internal::Throw_at_output_exception&)
+      {
+        // Sequentially write into the output iterator
+        for(std::size_t i=0; i<face_pairs.size(); ++i)
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:378]: there is something called `std::copy` :)

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h [hunk 2]
```diff
@@ -429,6 +544,15 @@ self_intersections(const FaceRange& face_range,
...
  *     \cgalParamExtra{The geometric traits class must be compatible with the vertex point type.}
  *   \cgalParamNEnd
+ *
+ *   \cgalParamNBegin{max_number}
+ *     \cgalParamDescription{the maximum number of self intersections that will be computed and returned by the function.}
+ *     \cgalParamType{unsigned int}
+ *     \cgalParamDefault{the number of self intersections in `tmesh`.}
+ *     \cgalParamExtra{In parallel mode, the number of returned self-intersections is at least `max_number`, but
+ *                     may be a little more, because some threads might keep running for a short time between the moment
+ *                     it is decided to stop and the moment they are actually stopped.}
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:554]:
```suggestion
 *     \cgalParamExtra{In parallel mode, the number of returned self-intersections is at least `max_number`(and not exactly that number) as for performance reason no strong synchronization is put on threads.}
```

File: BGL/include/CGAL/boost/graph/parameters_interface.h [hunk 3]
```diff
@@ -125,6 +125,7 @@ CGAL_add_named_parameter(allow_self_intersections_t, allow_self_intersections, a
 CGAL_add_named_parameter(non_manifold_feature_map_t, non_manifold_feature_map, non_manifold_feature_map)
 CGAL_add_named_parameter(polyhedral_envelope_epsilon_t, polyhedral_envelope_epsilon, polyhedral_envelope_epsilon)
 CGAL_add_named_parameter(face_epsilon_map_t, face_epsilon_map, face_epsilon_map)
+CGAL_add_named_parameter(max_number_t, max_number, max_number)
```
Comment (sloriot) [inline BGL/include/CGAL/boost/graph/parameters_interface.h:128]: max -> maximum in general for named parameters

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h [hunk 4]
```diff
@@ -227,6 +280,9 @@ self_intersections_impl(const FaceRange& face_range,
   VPM vpmap = choose_parameter(get_parameter(np, internal_np::vertex_point),
                                get_const_property_map(boost::vertex_point, tmesh));
 
+  const bool do_limit = !(is_default_parameter(get_parameter(np, internal_np::max_number)));
+  unsigned int max_number = choose_parameter(get_parameter(np, internal_np::max_number), 0);
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:284]:
```suggestion
  const unsigned int max_number = choose_parameter(get_parameter(np, internal_np::max_number), 0);
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h [hunk 5]
```diff
@@ -314,6 +402,24 @@ self_intersections_impl(const FaceRange& face_range,
 
   if(throw_on_SI)
     CGAL::box_self_intersection_d<CGAL::Sequential_tag>(box_ptr.begin(), box_ptr.end(), throwing_filter, cutoff);
+  else if(do_limit)
+  {
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:406]: in sequential mode a user could already write:
```
  std::size_t nbi=0;
  auto max_inter_counter = [&nbi, max_nb_reported](const std::pair<face_descriptor, face_descriptor>&)
  {
    if (++nbi == max_nb_reported)
      throw CGAL::internal::Throw_at_output_exception();
  };

  try{
    PMP::self_intersections<CGAL::Sequential_tag>(tm, boost::make_function_output_iterator(max_inter_counter));
  }
  catch (const CGAL::internal::Throw_at_output_exception&)
  {}
```
I'm surprised we require so many lines.

File: Box_intersection_d/include/CGAL/Box_intersection_d/segment_tree.h [hunk 6]
```diff
@@ -35,6 +35,13 @@ namespace CGAL {
 
 namespace Box_intersection_d {
 
+namespace internal {
```
Comment (sloriot) [inline Box_intersection_d/include/CGAL/Box_intersection_d/segment_tree.h:38]: does not seem used

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h [hunk 7]
```diff
@@ -290,14 +326,40 @@ self_intersections_impl(const FaceRange& face_range,
...
+        Throw_functor throwing_count_functor(atomic_counter, max_number, std::back_inserter(face_pairs));
+        Throwing_after_count_output_iterator count_filter(throwing_count_functor);
+        Filtered_intersecting_faces_filter limited_callback(tmesh, vpmap, gt, count_filter);
+        CGAL::box_self_intersection_d<ConcurrencyTag>(box_ptr.begin(), box_ptr.end(), limited_callback, cutoff);
+      }
+      catch(const CGAL::internal::Throw_at_output_exception&)
+      {
+        // Sequentially write into the output iterator
+        std::copy(face_pairs.begin(), face_pairs.end(), out);
+        //for(std::size_t i=0; i<face_pairs.size(); ++i)
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:352]: remove?

File: Box_intersection_d/include/CGAL/Box_intersection_d/segment_tree.h (hunk 6 above)
Comment (maxGimeno) [reply Box_intersection_d/include/CGAL/Box_intersection_d/segment_tree.h:38]: Not used, except in parallel mode you mean ?

File: BGL/test/BGL/test_cgal_bgl_named_params.cpp [hunk 8]
```diff
@@ -227,6 +228,7 @@ void test(const NamedParameters& np)
   check_same_type<60>(get_parameter(np, CGAL::internal_np::non_manifold_feature_map));
   check_same_type<61>(get_parameter(np, CGAL::internal_np::filter));
   check_same_type<62>(get_parameter(np, CGAL::internal_np::face_epsilon_map));
+  check_same_type<62>(get_parameter(np, CGAL::internal_np::maximum_number));
```
Comment (sloriot) [inline BGL/test/BGL/test_cgal_bgl_named_params.cpp:231]:
```suggestion
  check_same_type<68>(get_parameter(np, CGAL::internal_np::maximum_number));
```

**Comment by sloriot** [comment]
Successfully tested in [5.4-Ic-8](https://cgal.geometryfactory.com/CGAL/testsuite/results-5.4-Ic-8.shtml)

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h [hunk 9]
```diff
@@ -360,7 +440,17 @@ self_intersections_impl(const FaceRange& face_range,
  *     \cgalParamDefault{a \cgal Kernel deduced from the point type, using `CGAL::Kernel_traits`}
  *     \cgalParamExtra{The geometric traits class must be compatible with the vertex point type.}
  *   \cgalParamNEnd
+ *
+ *   \cgalParamNBegin{maximum_number}
+ *     \cgalParamDescription{the maximum number of self intersections that will be computed and returned by the function.}
+ *     \cgalParamType{unsigned int}
+ *     \cgalParamDefault{the number of self intersections in `face_range`.}
+ *     \cgalParamExtra{In parallel mode, the number of returned self-intersections is at least `maximum_number`
+ *     (and not exactly that number) as for performance reason no strong synchronization is put on threads.}
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:449]:
```suggestion
 *     (and not exactly that number) as no strong synchronization is put on threads for performance reasons.}
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h [hunk 10]
```diff
@@ -429,6 +519,14 @@ self_intersections(const FaceRange& face_range,
  *     \cgalParamDefault{a \cgal Kernel deduced from the point type, using `CGAL::Kernel_traits`}
  *     \cgalParamExtra{The geometric traits class must be compatible with the vertex point type.}
  *   \cgalParamNEnd
+ *
+ *   \cgalParamNBegin{maximum_number}
+ *     \cgalParamDescription{the maximum number of self intersections that will be computed and returned by the function.}
+ *     \cgalParamType{unsigned int}
+ *     \cgalParamDefault{the number of self intersections in `tmesh`.}
+ *     \cgalParamExtra{In parallel mode, the number of returned self-intersections is at least `maximum_number`
+ *     (and not exactly that number) as for performance reason no strong synchronization is put on threads.}
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:528]:
```suggestion
 *     (and not exactly that number) as no strong synchronization is put on threads for performance reasons.}
```

File: Polygon_mesh_processing/test/Polygon_mesh_processing/self_intersection_surface_mesh_test.cpp [hunk 11]
```diff
@@ -81,6 +81,53 @@ int test_self_intersections(const char* filename,
...
+template <typename K>
+int test_limited_self_intersections(const char* filename)
+{
+  typedef CGAL::Surface_mesh<typename K::Point_3>                Mesh;
+  typedef typename boost::graph_traits<Mesh>::face_descriptor    face_descriptor;
+
+  std::ifstream input(filename);
+  Mesh m;
+
+  if ( !input || !(input >> m) ) {
```
Comment (MaelRL) [inline Polygon_mesh_processing/test/Polygon_mesh_processing/self_intersection_surface_mesh_test.cpp:93]: All the IO efforts just to use old code... :>

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h [hunk 12]
```diff
@@ -16,7 +16,6 @@
 
 #ifndef CGAL_POLYGON_MESH_PROCESSING_SELF_INTERSECTIONS
 #define CGAL_POLYGON_MESH_PROCESSING_SELF_INTERSECTIONS
-
 #include <CGAL/license/Polygon_mesh_processing/predicate.h>
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:19]: Please restore that

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h [hunk 13]
```diff
@@ -360,7 +440,17 @@ self_intersections_impl(const FaceRange& face_range,
  *     \cgalParamDefault{a \cgal Kernel deduced from the point type, using `CGAL::Kernel_traits`}
  *     \cgalParamExtra{The geometric traits class must be compatible with the vertex point type.}
  *   \cgalParamNEnd
+ *
+ *   \cgalParamNBegin{maximum_number}
+ *     \cgalParamDescription{the maximum number of self intersections that will be computed and returned by the function.}
+ *     \cgalParamType{unsigned int}
+ *     \cgalParamDefault{the number of self intersections in `face_range`.}
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:447]: "No limit" would be clearer.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h [hunk 14]
```diff
@@ -360,7 +440,17 @@ self_intersections_impl(const FaceRange& face_range,
  *     \cgalParamDefault{a \cgal Kernel deduced from the point type, using `CGAL::Kernel_traits`}
  *     \cgalParamExtra{The geometric traits class must be compatible with the vertex point type.}
  *   \cgalParamNEnd
+ *
+ *   \cgalParamNBegin{maximum_number}
+ *     \cgalParamDescription{the maximum number of self intersections that will be computed and returned by the function.}
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:445]: I would say "detected" is a better fit than "computed"

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h (hunk 10 above)
Comment (MaelRL) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:528]: I would also remove parentheses.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h [hunk 15]
```diff
@@ -59,6 +58,32 @@ namespace CGAL {
...
+
+  std::atomic<unsigned int>& counter;
+  const unsigned int& max;
+
+  Output_iterator out;
+
+  Throw_at_count_reached_functor(std::atomic<unsigned int>& counter,
+                         const unsigned int& max,
+                         Output_iterator out)
+    : counter(counter), max(max), out(out)
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:72]:
```suggestion
  Throw_at_count_reached_functor(std::atomic<unsigned int>& counter,
                                 const unsigned int& max,
                                 Output_iterator out)
    : counter(counter), max(max), out(out)
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h [hunk 16]
```diff
@@ -252,7 +281,14 @@ self_intersections_impl(const FaceRange& face_range,
       if(throw_on_SI)
         throw CGAL::internal::Throw_at_output_exception();
       else
+      {
         *out++= std::make_pair(f, f);
+        ++counter;
+        if(do_limit && counter == maximum_number)
+        {
+          return out;
+        }
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/self_intersections.h:290]: Edge case, but if I pass `0` the current code will limit and dump a single self-intersection (same at the other place).

File: Polygon_mesh_processing/test/Polygon_mesh_processing/self_intersection_surface_mesh_test.cpp (hunk 11 above)
Comment (maxGimeno) [reply Polygon_mesh_processing/test/Polygon_mesh_processing/self_intersection_surface_mesh_test.cpp:93]: I blame the copy paste !

**Comment by lrineau** [comment]
@maxGimeno You should re-test this PR, modified by https://github.com/CGAL/cgal/pull/5837/commits/e83dc8775dc3747c823f4fc700a1d7c7b24f5984 since last test.

**Comment by sloriot** [comment]
Successfully tested in CGAL-5.4-Ic-28
